keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1080 @@
+get defined in separate KEPs. Developers will be able to add their own numeric
+models by recompiling kube-scheduler and/or CA without having to modify any of
+the upstream code.
+
+When opting into using numeric parameters for their DRA driver, driver
+developers no longer need to provide their own controller. Kubernetes takes
+over allocating and deallocating claims.
+
+Scheduling performance is expected to become better compared to using the
+PodSchedulingContext. However, supporting CA is the main reason for this KEP.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Enable cluster autoscaling when pods use resource claims, with correct
+  decisions and changing the cluster size by more than one node at a time.
+
+- Support node-local resources. Adding or removing nodes has no effect
+  on network-attached resources and therefore CA does not need to (and cannot)
+  simulate them.
+
+- Allow DRA driver developers to provide a user experience that is similar to
+  the one possible without numeric parameters. Ideally, users should not notice
+  at all that a driver is using numeric parameters under the hood.
+
+- Support adding and versioning multiple different numeric models without
+  having to modify the implementation of DRA or the numeric parameters
+  extension.
+
+- Keep the core DRA functionality from [KEP
+  #3063](https://github.com/kubernetes/enhancements/issues/3063) working as
+  defined there for Kubernetes 1.29. Cluster autoscaling is not always needed
+  and in-tree numeric model(s) are not expected to cover all use case, so the
+  flexibility offered by opaque parameters remains relevant.
+
+### Non-Goals
+
+- Support numeric class parameters. They remain opaque and can be used to
+  define configuration of resources, but must not affect where a claim can be
+  allocated.
+
+- Define exactly one numeric model that covers all current and future resources.
+  There cannot be a "one size fits all" solution.
+
+## Proposal
+
+### Publishing node capacity
+
+A numeric model defines a capacity type. The type has its own API group, kind
+and potentially different versions. It must support encoding and decoding JSON
+with an apimachinery scheme. The kubelet queries the registered DRA driver
+kubelet plugins for capacity information through a new gRPC call. Drivers
+respond by returning the JSON encoding of this capacity type. A streaming gRPC
+API allows drivers to update the capacity information.
+
+For example, a simple counter could be described as:
+```
+const GroupName = "counter.dra.config.k8s.io"
```

Comment:
I 'm not clear why this needs two names?  Maybe my own pre-assumptions are coloring it?  Is this like a Counter vs a Gauge in prometheus?

It seems like "counter.dra.config.k8s.io" expresses "this is a counter".  Why is it ""counter.dra.config.k8s.io/Capacity" and not ""counter.dra.config.k8s.io/Counter" or ""dra.config.k8s.io/Counter" ?

Reply:
This is literally just the API group. That API group then has different types for:
- publishing capacity ("capacity type")
- representing claim parameters ("parameter type")
- storing the result of an allocation ("allocation type")

We can name all of these types the same way in all numeric models (`Capacity`, `Parameters`, `Allocation`). In Go, the package name then becomes `counter.Capacity`. As a string, it would be `counter.dra.config.k8s.io/Capacity`. Either way, we don't need to name the type itself `Counter`.